click_odoo_contrib/update.py

```diff
@@ -21,6 +25,87 @@
+    max_td = timedelta(seconds=max_seconds)
+    own_pid_query = "SELECT pg_backend_pid()"
+    # SQL explained in https://wiki.postgresql.org/wiki/Lock_Monitoring
+    locks_query = """
+        SELECT
+            pg_stat_activity.datname,
+            pg_class.relname,
+            pg_locks.transactionid,
+            pg_locks.mode,
+            pg_locks.granted,
+            pg_stat_activity.usename,
+            pg_stat_activity.query,
+            pg_stat_activity.query_start,
+            AGE(NOW(), pg_stat_activity.query_start) AS age,
+            pg_stat_activity.pid
+        FROM
+            pg_stat_activity
+            JOIN pg_locks ON pg_locks.pid = pg_stat_activity.pid
+            JOIN pg_class ON pg_class.oid = pg_locks.relation
+        WHERE
+            NOT pg_locks.granted
+            AND pg_stat_activity.datname = %s
+        ORDER BY pg_stat_activity.query_start
+    """
+    # See https://stackoverflow.com/a/35319598/1468388
+    stop_query = "SELECT pg_cancel_backend(%s)"
+    if odoo.release.version_info < (9, 0):
+        params = {"dsn": odoo.sql_db.dsn(database)[1]}
+    else:
+        params = odoo.sql_db.connection_info_for(database)[1]
+    # Need a separate raw psycopg2 cursor without transactioning to avoid
+    # weird concurrency errors; this cursor will only trigger SELECTs, and
+    # it needs to access current Postgres server status, monitoring other
+    # transactions' status, so running inside a normal, transactioned,
+    # Odoo cursor would block such monitoring and, after all, offer no
+    # additional protection
+    with closing(psycopg2.connect(**params)) as watcher_conn:
+        watcher_conn.set_isolation_level(psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT)
+        while True:
+            # Wait some time before checking locks
+            sleep(beat)
+            # Ensure no long blocking queries happen
+            with closing(
+                watcher_conn.cursor(cursor_factory=psycopg2.extras.DictCursor)
+            ) as watcher_cr:
+                if _logger.level <= logging.DEBUG:
+                    watcher_cr.execute(own_pid_query)
+                    watcher_pid = watcher_cr.fetchone()[0]
+                    _logger.debug(
+                        "DB lock watcher running with postgres PID %d", watcher_pid
+                    )
+                watcher_cr.execute(locks_query, (database,))
+                locks = watcher_cr.fetchall()
+                if locks:
+                    _logger.warning("%d locked queries found", len(locks))
+                    _logger.info("Query details: %r", locks)
+                for row in locks:
+                    if row["age"] > max_td:
+                        _logger.warning(
+                            "Long locked query detected; aborting update cursor..."
+                        )
+                        # Terminate the query to abort the parallel update
+                        watcher_cr.execute(stop_query, (row["pid"],))
```

Comment:
Is it certain that killing row["pid"] will stop the update process and not the actual Odoo instance?